### ISSUE TITLE
about bm(backward mapping)

### ISSUE THREAD

**Issue by jeffer0323** [keepfoolisher/My-DocTr-Plus#2]
Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?

**Comment by keepfoolisher** [comment]
> Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?

I don't have permission to download Doc3d，I just make some fake data myself, i can not share my data until now because its used in my project, but i can share some personal opinion，origin image and bm file  is one-to-one correspondence，you can crop a region, and then find the correspond area in bm file, You can definitely achieve it yourself.

**Comment by jeffer0323** [comment]
> > Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?
> 
> I don't have permission to download Doc3d，I just make some fake data myself, i can not share my data until now because its used in my project, but i can share some personal opinion，origin image and bm file is one-to-one correspondence，you can crop a region, and then find the correspond area in bm file, You can definitely achieve it yourself.

Yes, I see that your model preserves the background content after inference, which is different from the original paper's approach. However, I personally think your approach is better, which is why I wanted to know how you achieved the cropping. Finally, thank you for your suggestion, I will try it out.

**Comment by BigTree-git** [comment]
can you share doc3d data For me？ I don't have permission to download Doc3d

**Comment by songzr0** [comment]
> > Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?
> 
> I don't have permission to download Doc3d，I just make some fake data myself, i can not share my data until now because its used in my project, but i can share some personal opinion，origin image and bm file is one-to-one correspondence，you can crop a region, and then find the correspond area in bm file, You can definitely achieve it yourself.

您做了很优秀的工作！我遇到一个问题：对origin image及bm做裁剪后，经由opencv的reamap进行扭曲校正，得到的校正图片相比于crop origin image会缺失边缘。但是看您训练出的模型，校正图片包含全部crop origin image信息。请问您是怎么做的呢？

**Comment by jeffer0323** [comment]
> can you share doc3d data For me？ I don't have permission to download Doc3d

The dataset is very large, and I did not use cloud storage. You can use an edu email to fill out the form to apply for download permission. The form link is: https://forms.gle/RTfi7LUSrt891VuN8. That's how I downloaded it.

**Comment by jeffer0323** [comment]
@songzr0 我实验了两种方案：一、裁剪uv file ， 通过裁剪后的uv计算得到bm，uv是3通道文件，第一个通道取值[0 , 1],0代表是背景，1代表文档，想要裁剪后保留背景，需要重新计算文档区域的uv坐标。 具体分两步，第一步是裁剪后对uv以最大值和最小值为范围重新做归一化，第二步是文档区域uv根据背景区域距离进行位移操作（可能描述的不太准确）；二、裁剪时只保留文档区域，对应裁剪uv，训练后就能有效推理含背景的文档图像（但对大区域背景的文档图像效果一般）。以上，仅供参考。

**Comment by jeffer0323** [comment]
@songzr0 顺便说下，我觉得模型推理后产生多余的黑边和整个文档格格不入，所以将其替换成白色或者采用最近邻插值得到调整后的文档，以便下游任务的处理。希望能对你有所帮助！

**Comment by BigTree-git** [comment]
训练几个batch后，我在这个地方输出flow_up值，显示为nan，入下图所示位置：
![image](https://github.com/keepfoolisher/My-DocTr-Plus/assets/29195979/592c792d-5569-4b48-950a-2bb38c34755d)
请问你遇到类似的情况吗？

**Comment by BigTree-git** [comment]
补充：
![image](https://github.com/keepfoolisher/My-DocTr-Plus/assets/29195979/8692af25-d0c2-4782-93fe-f33b12ff8b8f)

**Comment by songzr0** [comment]
> @songzr0 我实验了两种方案：一、裁剪uv file ， 通过裁剪后的uv计算得到bm，uv是3通道文件，第一个通道取值[0 , 1],0代表是背景，1代表文档，想要裁剪后保留背景，需要重新计算文档区域的uv坐标。 具体分两步，第一步是裁剪后对uv以最大值和最小值为范围重新做归一化，第二步是文档区域uv根据背景区域距离进行位移操作（可能描述的不太准确）；二、裁剪时只保留文档区域，对应裁剪uv，训练后就能有效推理含背景的文档图像（但对大区域背景的文档图像效果一般）。以上，仅供参考。

谢谢您的回复，听起来很有用，但由于我没有接触过3d图像的处理，uv坐标转换对我来说很困难。请问您上面所属或模型训练的代码您能否发给我一份学习一下？如果可以，我不胜感激。我的邮箱是837927569@qq.com.

**Comment by songzr0** [comment]
> @songzr0 我实验了两种方案：一、裁剪uv file ， 通过裁剪后的uv计算得到bm，uv是3通道文件，第一个通道取值[0 , 1],0代表是背景，1代表文档，想要裁剪后保留背景，需要重新计算文档区域的uv坐标。 具体分两步，第一步是裁剪后对uv以最大值和最小值为范围重新做归一化，第二步是文档区域uv根据背景区域距离进行位移操作（可能描述的不太准确）；二、裁剪时只保留文档区域，对应裁剪uv，训练后就能有效推理含背景的文档图像（但对大区域背景的文档图像效果一般）。以上，仅供参考。

您好，我找到了uv坐标转bm的脚本。但我有两个新的问题。doc3d数据集中的图片包含三部分，杂乱背景，文档底纹，文档中的文字。背景区域的uv值全部为0，如何得到含背景位移的bm？换言之，若对img进行裁剪，能得到含有部分杂乱背景+文档的图片，但对应的uv图和bm图仍旧只包含前景文档？

**Comment by songzr0** [comment]
> @songzr0 我实验了两种方案：一、裁剪uv file ， 通过裁剪后的uv计算得到bm，uv是3通道文件，第一个通道取值[0 , 1],0代表是背景，1代表文档，想要裁剪后保留背景，需要重新计算文档区域的uv坐标。 具体分两步，第一步是裁剪后对uv以最大值和最小值为范围重新做归一化，第二步是文档区域uv根据背景区域距离进行位移操作（可能描述的不太准确）；二、裁剪时只保留文档区域，对应裁剪uv，训练后就能有效推理含背景的文档图像（但对大区域背景的文档图像效果一般）。以上，仅供参考。

您好，我找到了uv坐标转bm的脚本。但我有两个新的问题。
1. doc3d数据集中的图片包含三部分，杂乱背景，文档底纹，文档中的文字。背景区域的uv值全部为0，如何得到含背景位移的bm？换言之，若对img进行裁剪，能得到含有文档+部分杂乱背景的图片，但对应的uv图和bm图仍旧只包含前景文档？
2.我没理解您所述两种方案的区别，能否解释一下？

**Comment by huycq1712** [comment]
> > Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?
> 
> I don't have permission to download Doc3d，I just make some fake data myself, i can not share my data until now because its used in my project, but i can share some personal opinion，origin image and bm file is one-to-one correspondence，you can crop a region, and then find the correspond area in bm file, You can definitely achieve it yourself.

@keepfoolisher 
Thank you for your great work!
I want to make my own dataset like doc3d, but i dont know how to get the .obj files to build data from flatten images. Can you share these .obj files for me?

**Comment by jeffer0323** [comment]
@huycq1712 
There is no complete .obj file in the original Doc3D dataset, and I have submitted an application to the original author, but I have not received a response. Fortunately, I found a script for creating datasets using blender software at https://github.com/FelixHertlein/inv3d-generator. The disadvantage is that there are few .obj files in the script. In view of this, I spent two days learning the usage of blender and created a large number of meshes (.obj files) using blender. You can also create your own large dataset in this way. I wish you success!

**Comment by chenjjcccc** [comment]
> Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?

我最近也在做相关的工作，可以加个好友一起分享一下吗～

**Comment by pineking** [comment]
> > Great job on the paper replication! but I have a question, In the original Doc3d dataset, the bm file size is 448x448x2. How can we crop it to 288x288x2?
> 
> 我最近也在做相关的工作，可以加个好友一起分享一下吗～

我最近也在看类似的工作，我微信号 pineking

**Comment by bleachyin** [comment]
> @songzr0 我实验了两种方案：一、裁剪uv file ， 通过裁剪后的uv计算得到bm，uv是3通道文件，第一个通道取值[0 , 1],0代表是背景，1代表文档，想要裁剪后保留背景，需要重新计算文档区域的uv坐标。 具体分两步，第一步是裁剪后对uv以最大值和最小值为范围重新做归一化，第二步是文档区域uv根据背景区域距离进行位移操作（可能描述的不太准确）；二、裁剪时只保留文档区域，对应裁剪uv，训练后就能有效推理含背景的文档图像（但对大区域背景的文档图像效果一般）。以上，仅供参考。

大神好，请教个问题，我更倾向于用第二种方法，但是裁剪uv区域后，bm是如何重新计算的？
